vassal-app/src/main/java/VASSAL/build/module/map/MenuDisplayer.java

```diff
@@ -267,6 +270,16 @@ public void popupMenuWillBecomeVisible(PopupMenuEvent evt) {
 
     // It is possible for the map to close before the menu is displayed
     if (map.getView().isShowing()) {
+
+      // Inform the piece where player clicked, if it wants to know.  
+      final ChangeTracker ch = new ChangeTracker(p);
+      p.setProperty("ClickedX", e.getPoint().x);
+      p.setProperty("ClickedY", e.getPoint().y);
+      Command comm = ch.getChangeCommand();
+      if ((comm != null) && !comm.isNull()) {
+        GameModule.getGameModule().sendAndLog(comm);
+      }      
+
```

Comment:
Hi Brian,

Sorry, I am only getting to review this now. This change as implemented has a few problems.

1. If the designer does not create Dynamic Properties ClickedX and ClickedY, the change still appears to work because ClickedX and ClickedY will end up stored in the scratch-pad property HashMap in BasicPiece. However, you end up with all the same problems as the backXXX variables used to have:- Not sent to clients or log or save files, undo does not work.

Because it appears to work on a single client, the designer may 'forget' or decide not to create the Dynamic Properties.

As an alternative, you could make ClickedX and ClickedY Persistent properties using the interface I created to fix SendToLocation by doing something like this:

Command comm  = new NullCommand();
comm = setPersistentProperty(BasicPiece,CLICKED_X, String.valueOf(e.getPoint().x));
comm.append (setPersistentProperty(BasicPiece,CLICKED_Y, String.valueOf(e.getPoint().y));
if ((comm != null) && !comm.isNull()) {

2. However, there is a deeper problem in that whether using user supplied Dynamic Properties or persistent properties, this solution generates a Command each time the player right-clicks on a piece, regardless of whether or not they actually select a menu option. 

When you replay the log, there will appear to be empty steps to click through. 

Even if you do select a Command, this will generate 2 steps, an apparently empty one when the Clicked props are set, followed by the actual menu selection execution. Currently, a single menu selection only ever generate a single log step. Now it will generate 2.

The Command generated by the KeyCommand selected on the menu needs to be appended to the Command to set the properties,but only after the menu item is selected.
 
This one is tricky because the individual KeyCommands in the popup handle their own activation directly through KeyCommand.actionPerformed().  

You may be able to mediate this through the KeyBuffer. Set the ClickedX and ClickedY into the KeyBuffer in MenuDisplayer, then use them in KeyCommand.actionPerformed() which already accesses the KeyBuffer to get the list of currently selected units.

There are cases where actionPerformed is call directly on KeyCommands not as a result of the MenuDisplayer, so you will also need to determine that the actionPerformed is actually a result of a MenuDisplayer selection. Tricky, but I suspect the KeyBuffer will only have units in it at MenuDisplayer time.

Reply:
Ooooo. Didn't know about the scratchpad property. Is there a way to check definitively for the existence of the DynamicProperty in advance of sending the SetProperty?

Reply:
Not really, because the same Command may be applied to a whole Stack by KeyBuffer. On an individual Counter, you would need to run down the trait stack and check each one. Just change them to Persistent properties, that solves the problems and is simpler for the designer. No need to add 2 more traits to the stack just to access ClickedX/Y.

Reply:
I would definitely approach it by 
1. adding clickPoint to KeyBuffer and have menuDisplayer set it.
2. Modify KeyBuffer.keyCommand to set the Persistent properties from clickPoint, then append p.keyEvent(stroke) for each piece. The handles the case where a menu command is applied to a selection of units.

I think that's all you need. All Keycommands issued from MenuDisplayer (and only KeyCommand issued from the Menudisplayer) have global set to true, so you can do all the work in KeyBuffer

Reply:
Please check #156 -- it seems to be running correctly but give it a once-over.